kie-server-parent/kie-server-remote/kie-server-client/src/main/java/org/kie/server/client/jms/AsyncResponseHandler.java

```diff
@@ -0,0 +1,162 @@
+        return new ServiceResponsesList(Arrays.asList(messageSentResponse));
+    }
+
+    @Override
+    public void dispose(Connection connection, Session session) {
+        // no op as the resources are closed from within message listener (via separate thread)
+    }
+
+    private class AsyncMessageListener implements MessageListener  {
+        private String selector;
+        private MessageConsumer consumer;
+        private Marshaller marshaller;
+        private KieServicesClient owner;
+
+        private Connection connection;
+        private Session session;
+
+        public AsyncMessageListener(Connection connection, Session session, String selector, MessageConsumer consumer, Marshaller marshaller, KieServicesClient owner) {
+            this.selector = selector;
+            this.consumer = consumer;
+            this.marshaller = marshaller;
+            this.owner = owner;
+            this.connection = connection;
+            this.session = session;
+        }
+
+        @Override
+        public void onMessage(Message message) {
+            try {
+                ((KieServicesClientImpl) owner).setConversationId(message.getStringProperty(JMSConstants.CONVERSATION_ID_PROPERTY_NAME));
+
+                String responseStr = ((TextMessage) message).getText();
+                logger.debug("Received response from server '{}'", responseStr);
+
+                ServiceResponsesList cmdResponse = marshaller.unmarshall(responseStr, ServiceResponsesList.class);
+                logger.debug("Unmarshalled response from async delivery {} calling callback {}", cmdResponse, callback);
+
+                callback.onResponse(selector, cmdResponse);
+                logger.debug("Callback {} successfully invoked with response {}", callback, cmdResponse);
+            } catch (Exception e) {
+                logger.error("Error while receiving message due to {}, this means response from the server won't be delivered to client", e.getMessage(), e);
+            } finally {
+                if (consumer != null) {
+                    try {
+                        consumer.close();
+                    } catch (JMSException e) {
+                        logger.warn("Error when closing JMS consumer due to {}", e.getMessage());
+                    }
+                }
+                // submit work to executor service to close resources
+                // as they cannot be closed from message listener
+                // due to AMQ129006: It is illegal to call this method (session.close()) from within a Message Listener
+                executorService.submit(new Runnable() {
+                    @Override
+                    public void run() {
+                        try {
+                            if (session != null) {
+                                session.close();
+                                logger.debug("Session closed via separate thread.");
```

Comment:
Could we perhaps add more info about the session in to the log message? From my experience the more specific data the logging message contain the better, it is then much easier to figure our what is going on.